### PR TITLE
Attackby code will now default to constructing source with target when both exactly match a recipe.

### PR DESCRIPTION
**Pull request by MrsTonedOne**
I'm 99% sure this broke something, but holy shit was this really so fucking hard?

I did this in 15 minutes in the github editor.

I'll fix what ever bugs with this are found later, its gonna get feature frozen anyway

@tgstation/commit-access

### CODE REVIEW

**Comment by Core0verload** [comment]
I'm 101% sure this broke something.

File: code/game/objects/items.dm [hunk 1]
```diff
@@ -345,6 +345,29 @@ var/global/image/fire_overlay = image("icon" = 'icons/effects/fire.dmi', "icon_s
 
 			else if(S.can_be_inserted(src))
 				S.handle_item_insertion(src)
+		return 
+	if (istype(user,/mob/living/carbon/human))
```
Comment (ChangelingRain) [inline code/game/objects/items.dm]: use ishuman(user)

**Comment by Cheridan** [comment]
>a ton a loops every time an item is hit with something else

wew

File: code/game/objects/items.dm (hunk 1 above)
Comment (RemieRichards) [inline code/game/objects/items.dm]: No, cos humans aren't the only people who can craft.

Comment (ChangelingRain) [inline code/game/objects/items.dm]: This is true, but that was referring to this specific check in its current state.

Comment (RemieRichards) [inline code/game/objects/items.dm]: ya, and it's current state is wrong :P

**Comment by Razharas** [comment]
Only  solution is a global list of lists in that is being filled with type-keys to list that holds all recipes that have this type as req, on new() of each item the value of under the key-type is fetched into an item's var/list/self_recipes and on all attacks only this small list(usually empty one) is checked
Global list is is updated per recipes' new() call, where it loops over reqs and adds self to the list under the iterator type key

File: code/game/objects/items.dm (hunk 1 above)
Comment (MrStonedOne) [inline code/game/objects/items.dm]: human is the type that has the personal crafting datum attached to it that i use below.

**Comment by MrStonedOne** [comment]
cheri we do 10,000 loops every half a second in atmos just fine, shit gets hit far less often.

File: code/game/objects/items.dm (hunk 1 above)
Comment (RemieRichards) [inline code/game/objects/items.dm]: @MrStonedOne 

```
/mob/living/simple_animal/drone/OpenCraftingMenu()
	handcrafting.ui_interact(src)
```
?

**Comment by optimumtact** [comment]
I dont' understand what the point of this pr is, can someone explain what is going on.

File: code/game/objects/items.dm (hunk 1 above)
Comment (MrStonedOne) [inline code/game/objects/items.dm]: https://github.com/tgstation/tgstation/blob/4bc5e1d085de4f89b428515a8fdc6cf2823198f4/code/modules/mob/living/carbon/human/human_defines.dm#L57

**Comment by MrStonedOne** [comment]
then automates slap crafting into datum crafting by looking for matching datum crafting recipies during any uncaught attackby.

File: code/game/objects/items.dm (hunk 1 above)
Comment (RemieRichards) [inline code/game/objects/items.dm]: https://github.com/tgstation/tgstation/blob/4bc5e1d085de4f89b428515a8fdc6cf2823198f4/code/modules/mob/living/simple_animal/friendly/drone/_drone.dm#L64

Can you just stop being a tool.

Comment (MrStonedOne) [inline code/game/objects/items.dm]: Explain to me how i can make this work without having to make copy pasted code, one for drone's type, one for human's type.

and that alone is untenable, its snowflake and would rightfully be declined by the maintainers.

**Comment by LatD** [comment]
I prefer slap crafting over personal crafting too so I hope this works. Just some questions about this new system.

Does it matter which item are you trying to use if you need to combine two?
What happens with items that need to be set to attach-mode before putting them together? Like igniters, welding tools etc. I don't even know how tablecrafting/personal crafting uses those.
Does this work properly with items that have middle steps? Like flamethrower and spears and some other items.

**Comment by optimumtact** [comment]
I've thought a lot about how to do the slap crafting properly, I don't think tablecrafting's recipe approach is quite the right fit. 

A proper solution has the following
1) Tooltype flags to allow any obj/item to be registered as a "tool", along with either vars, or a tooltype datum to handle the following, sound on use, message on use, resource usage, can_use_tool(mob)
2) ability to customise the message output for each construction step (state change)
3) construction state datum that follows the object through construction (including across objects if the recipe creates a new item and deletes the old one, tracks currrent recipe and progression through it)
4) ability to modify speed of a given step by recipe developer
5) ability to change resource use of a given step by recipe developer
6) each recipe step needs to define an end state value and then any recipe can configure that as an available starting state, next tool use defines which new recipe starts to be followed, as many recipes for slapcrafting have branching based on the tool used.
7) each typeclass/tooltype combination knows about the list of recipes valid for it, so only has to review those on attack_by to reduce performance overhead
8)generic recipe steps to be reused in recipes with slight modification i.e step_useTool(tooltype)

etc etc

**Comment by Razharas** [comment]
Midsteps are the worst part about crafting
Tablecrafting was possible to do exactly because it lacks midteps
All our crafting attempts either lack proper midsteps like machine creation or are a horrible, hardcoded mess where each recipe vary from another by 1 step, like computer construction
Doing a generic system of crafting with fully adjustable midsteps will be much harder than any crafting system we have currently
Basically such a crafting system will be a small language designed to define recipes, and we all know what happens with small language systems

**Comment by AnturK** [comment]
Some kind of DSL for these would probably be the best.

**Comment by optimumtact** [comment]
This is not the right approach
